game-engine/src/physics/collision.rs

```diff
@@ -0,0 +1,376 @@
+pub fn pop_coliders(
+    normal_distance: Vec3,
+    t1: &mut Transform,
+    t2: &mut Transform,
+    rb1: &RidgidBody,
+    rb2: &RidgidBody,
+) {
+    //debug_assert!(normal_distance.magnitude_squared() > 0.0); //TODO
+    // cant move static coliders
+    if rb1.is_static && rb2.is_static {
+        return;
+    }
+
+    const POP_SIZE: f32 = 1.1;
+    let pop = normal_distance * POP_SIZE;
+    if rb1.is_static {
+        t2.position -= pop;
+    } else if rb2.is_static {
+        t1.position += pop;
+    } else {
+        t2.position -= pop * 0.5;
+        t1.position += pop * 0.5;
+    }
+}
+
+pub fn solve_colliding(
+    c1: &Collider,
+    rb1: &mut RidgidBody,
+    t1: &mut Transform,
+    c2: &Collider,
+    rb2: &mut RidgidBody,
+    t2: &mut Transform,
+) {
+    let w1 = get_position(t1, c1);
+    let w2 = get_position(t2, c2);
+
+    debug_assert_finite!(w1);
+    debug_assert_finite!(w2);
+
+    match c1 {
+        Collider::SphereColider(b1) => match c2 {
+            Collider::SphereColider(b2) => {
+                collide_sphere_vs_sphere(b1, rb1, t1, w1, b2, rb2, t2, w2)
+            }
+            Collider::BoxColider(b2) => collide_sphere_vs_box(b1, rb1, t1, w1, b2, rb2, t2, w2),
+        },
+        Collider::BoxColider(b1) => match c2 {
+            Collider::SphereColider(b2) => {
+                collide_sphere_vs_box(b2, rb2, t2, w2, b1, rb1, t1, w1)
+            }
+            Collider::BoxColider(b2) => collide_box_vs_box(b1, rb1, t1, w1, b2, rb2, t2, w2),
+        },
+    }
+}
+
+pub fn collide(
+    rb1: &mut RidgidBody,
+    t1: &mut Transform,
+    cc1: &Vec<Collider>,
+    t2: &mut Transform,
+    rb2: &mut RidgidBody,
+    cc2: &Vec<Collider>,
+) -> bool {
+    let mut has_colided = false;
+
+    for c1 in cc1 {
+        for c2 in cc2 {
+            if is_colliding(c1, t1, c2, t2) {
+                has_colided = true;
+
+                let rot_1 = Mat3::from(t1.rotation);
+                let rot_2 = Mat3::from(t2.rotation);
+
+                let i_1 = rot_1 * c1.inv_inertia_tensor() * rot_1.transposed();
+                let i_2 = rot_2 * c2.inv_inertia_tensor() * rot_2.transposed();
+
+                debug_assert_finite!(rb1.velocity());
+                debug_assert_finite!(rb2.velocity());
+                debug_assert_finite!(rb1.angular_velocity(i_1));
+                debug_assert_finite!(rb2.angular_velocity(i_2));
+
+                //rb1.is_active = true;
+                //rb2.is_active = true;
+                rb1.is_active_time = 0.0;
+                rb2.is_active_time = 0.0;
+
+                solve_colliding(c1, rb1, t1, c2, rb2, t2);
+            }
+        }
+    }
+
+    rb1.is_colliding_this_frame = has_colided || rb1.is_colliding_this_frame;
+    rb2.is_colliding_this_frame = has_colided || rb2.is_colliding_this_frame;
+
+    has_colided
+}
+
+impl RidgidBody {
+    pub fn add_impulse(&mut self, force: Vec3) {
```

Comment:
Weird naming, should be `add_force(..., force: ...` or `add_impulse(..., impulse: ...`